src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+                            //This case handle additions
+                            originalObj = Activator.CreateInstance(changedObj.ExpectedClrType);
+                            original.Add(originalObj);
+                        }
+
+                        //Patch for addition/update. This will call Delta<T> for each item in the collection
+                        changedObj.Patch(originalObj);
+                    }
+
+                    edmChangedObject = changedObj;
+                }
+                catch
+                {
+                    //For handling the failed operations.
+                    HandleFailedOperation(changedObj, operation, ref edmChangedObject, originalObj);
+                }
+
+                edmChangedObjectCollection.Add(edmChangedObject);
+            }
+
+            return edmChangedObjectCollection;
+        }
+               
+
+        private void HandleFailedOperation(dynamic changedObj, DataModificationOperationKind operation, ref IEdmChangedObject edmChangedObject, TStructuralType originalObj)
+        {
+            DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+
+            // This handles the Data Modification exception. This adds Core.DataModificationException annotation and also copy other instance annotations.
+            //The failed operation will be based on the protocol
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    edmChangedObject = changedObj;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        Type type = typeof(EdmDeltaDeletedEntityObject<>).MakeGenericType(changedObj.ExpectedClrType);
+                        EdmDeltaDeletedEntityObject edmDeletedObject  = Activator.CreateInstance(type, EntityType) as EdmDeltaDeletedEntityObject;
+                        object contentId = changedObj.TransientInstanceAnnotationContainer.GetResourceAnnotation("Core.ContentID");
+                        if(contentId != null)
+                        {
+                            edmDeletedObject.Id = contentId.ToString();                            
+                        }
+                        else
+                        {
+                            edmDeletedObject.Id = string.Empty;
+                        }
+
+                        CopyInstanceAnnotations(changedObj, edmDeletedObject);
+                        edmDeletedObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);                       
+                        edmChangedObject = edmDeletedObject;
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                case DataModificationOperationKind.Unlink:
```

Comment:
>Unlink [](start = 51, length = 6)

Unlink would generally be used with a DeltaDeletedLink, which is different from Deleted with Reason=Removed.  I don't see handling of addedLink and deletedLink.